File: Backend/app/models/activity.py
"""
Activity Framework - Backend Pydantic models
Maps to TypeScript ActivityItem interface
"""

from typing import Optional, List, Literal
from pydantic import BaseModel

# Type definitions
ActivityType = Literal[
    "image_to_word",
    "one_step_instruction",
    "two_step_sequence",
    "counting",
    "visual_addition",
    "pattern",
    "comparison",
    "logic_choice",
    "sequence_ordering",
    "focus_filter"
]

ModuleId = Literal["M1", "M2", "M3"]
LessonId = Literal["1.1", "1.2", "1.3", "2.1", "2.2", "2.3", "3.1", "3.2"]
Difficulty = Literal["easy", "medium", "hard"]
NeuroType = Literal["Dyslexia", "ADHD", "ASD"]


class ActivityOption(BaseModel):
    """Single choice/answer option"""
    id: str
    label: str
    isCorrect: bool
    ttsText: Optional[str] = None
    imageUrl: Optional[str] = None
    imageAlt: Optional[str] = None
    tags: Optional[List[str]] = None


class ActivityAccessibility(BaseModel):
    """Accessibility metadata for neurodiversity support"""
    recommendedFor: List[NeuroType]
    enableTtsOnHover: Optional[bool] = False
    showProgressBar: Optional[bool] = False
    avoidMetaphors: Optional[bool] = True
    consistentFeedback: Optional[bool] = True


class ActivityItem(BaseModel):
    """Core schema for each question/task"""
    id: str
    moduleId: ModuleId
    lessonId: LessonId
    type: ActivityType

    # What the student sees / hears
    instruction: str
    instructionTts: Optional[str] = None
    stimulusImageUrl: Optional[str] = None
    stimulusImageAlt: Optional[str] = None
    stimulusEmoji: Optional[str] = None
    stimulusDescription: Optional[str] = None

    # For multi-step sequences
    steps: Optional[List[str]] = None

    options: List[ActivityOption]

    # Difficulty & tracking
    difficulty: Difficulty
    maxTimeSeconds: Optional[int] = None
    targetCategory: Optional[str] = None

    # Accessibility / neurodiversity hints for UI
    accessibility: ActivityAccessibility

    class Config:
        json_schema_extra = {
            "example": {
                "id": "M1_L1_Q1",
                "moduleId": "M1",
                "lessonId": "1.1",
                "type": "image_to_word",
                "instruction": "Match the picture to the correct word.",
                "instructionTts": "Look at the picture and click the word that matches.",
                "stimulusImageUrl": "/images/module1/lesson1/pig.png",
                "stimulusImageAlt": "A pink pig standing on grass",
                "stimulusDescription": "A pink pig",
                "difficulty": "easy",
                "options": [
                    {
                        "id": "A",
                        "label": "Pig",
                        "isCorrect": True,
                        "ttsText": "pig"
                    },
                    {
                        "id": "B",
                        "label": "Big",
                        "isCorrect": False,
                        "ttsText": "big"
                    }
                ],
                "accessibility": {
                    "recommendedFor": ["Dyslexia"],
                    "enableTtsOnHover": True,
                    "showProgressBar": False,
                    "avoidMetaphors": True,
                    "consistentFeedback": True
                }
            }
        }

File: Frontend/src/types/activity.ts
/**
 * Activity Framework - Consistent structure for all activity types
 * Supports: image, text, emoji, sequence activities
 * Includes TTS/audio support and neurodiversity-specific tweaks
 */

export type ActivityType =
  | "image_to_word"          // Lesson 1.1: Match picture to word
  | "one_step_instruction"   // Lesson 1.2: Single instruction
  | "two_step_sequence"      // Lesson 1.3: Two-step sequence
  | "counting"               // Lesson 2.1: Count objects
  | "visual_addition"        // Lesson 2.2: Visual math
  | "pattern"                // Lesson 2.3: Patterns
  | "comparison"             // Lesson 2.3: Comparisons
  | "logic_choice"           // Lesson 2.3: Logic choices
  | "sequence_ordering"      // Lesson 3.1: Drag & drop sequences
  | "focus_filter";          // Lesson 3.2: Focus tasks (click targets)

export type ModuleId = "M1" | "M2" | "M3";
export type LessonId = "1.1" | "1.2" | "1.3" | "2.1" | "2.2" | "2.3" | "3.1" | "3.2";
export type Difficulty = "easy" | "medium" | "hard";
export type NeuroType = "Dyslexia" | "ADHD" | "ASD";

/**
 * Activity Option - represents a single choice/answer option
 */
export interface ActivityOption {
  id: string;              // "A", "B", "opt1", etc.
  label: string;           // What we show (text or emoji, like "Pig" or "🦋 🦋")
  isCorrect: boolean;      // true/false
  ttsText?: string;        // What TTS should say (plain text)
  imageUrl?: string;       // If option is an image
  imageAlt?: string;       // Description of the image for screen readers/TTS
  tags?: string[];         // e.g. ["target", "distractor", "bird", "green"]
}

/**
 * Accessibility metadata for neurodiversity support
 */
export interface ActivityAccessibility {
  recommendedFor: NeuroType[];
  enableTtsOnHover?: boolean;     // e.g. for Dyslexia word tasks
  showProgressBar?: boolean;      // e.g. for ADHD focus sprint
  avoidMetaphors?: boolean;       // Always true for ASD
  consistentFeedback?: boolean;   // Always true for ASD modules
}

/**
 * Activity Item - Core schema for each question/task
 */
export interface ActivityItem {
  id: string;                 // "M1_L1_Q1"
  moduleId: ModuleId;
  lessonId: LessonId;
  type: ActivityType;

  // What the student sees / hears
  instruction: string;        // Onscreen text: "Click the sad boy."
  instructionTts?: string;    // Spoken version (can be simpler)
  stimulusImageUrl?: string;  // If there is a main picture
  stimulusImageAlt?: string; // What that image means in text
  stimulusEmoji?: string;     // For emoji-only stimuli if needed
  stimulusDescription?: string; // Human-readable: "a pink pig", "three pizzas"

  // For multi-step sequences
  steps?: string[];           // ["First click the Spoon", "then click the Fork"]

  options: ActivityOption[];  // Choices, with correct marking

  // Difficulty & tracking
  difficulty: Difficulty;
  maxTimeSeconds?: number;    // For timed tasks like focus sprint
  targetCategory?: string;    // e.g. "Green Things", "Birds" (for focus tasks)

  // Accessibility / neurodiversity hints for UI
  accessibility: ActivityAccessibility;
}

/**
 * Helper type for backward compatibility with existing Activity interface
 */
export interface Activity {
  activityId: string;
  topic: string;
  difficulty: Difficulty | string;
  modality: "text" | "audio" | "visual" | string;
  question: string;
  passage?: string;
  options?: string[];
}

